### PR TITLE
pkcs11-provider: init at 0.3

### PR DESCRIPTION
**Pull request by numinit**
## Description of changes

Init pkcs-provider at 0.3. This is a PKCS#11 provider for OpenSSL 3.

Currently this is packaged in https://github.com/numinit/nixpkcs, pushing upstream.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/pk/pkcs11-provider/package.nix [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  postPatch = ''
+    patchShebangs .
+
+    # Makefile redirects to logfiles; make sure we can catch them.
+    for name in softokn softhsm; do
+      ln -s /dev/stderr tests/setup-$name.log
+    done
+  '';
+
+  preConfigure = "autoreconf -fi";
```
Comment (Aleksanaa) [inline pkgs/by-name/pk/pkcs11-provider/package.nix:33]: Have you tried `pkgs.autoreconfHook`?

File: pkgs/by-name/pk/pkcs11-provider/package.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  enableParallelBuilding = true;
+  enableParallelInstalling = false;
```
Comment (Aleksanaa) [inline pkgs/by-name/pk/pkcs11-provider/package.nix:36]: Is this not reproducible?

Comment (numinit) [reply pkgs/by-name/pk/pkcs11-provider/package.nix:36]: I'll add a comment; it frequently dies due to a race.

File: pkgs/by-name/pk/pkcs11-provider/package.nix (hunk 1 above)
Comment (numinit) [reply pkgs/by-name/pk/pkcs11-provider/package.nix:33]: Didn't have any idea this existed, good call!

File: pkgs/by-name/pk/pkcs11-provider/package.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
+{ lib, stdenv, fetchFromGitHub, runCommandLocal
+, openssl, nss, p11-kit
+, opensc, gnutls, expect
+, autoconf, autoconf-archive, automake, pkg-config, libtool
+}:
+
+stdenv.mkDerivation rec {
+  pname = "pkcs11-provider";
```
[the remark this answers is not in the document]
Comment (numinit) [reply pkgs/by-name/pk/pkcs11-provider/package.nix:8]: Yeah I stuck with the original project name for congruency with [pkcs11helper](https://search.nixos.org/packages?channel=23.11&size=50&sort=relevance&type=packages&query=pkcs11) and similar.

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**